tests: eval and build more tests

jopejoe1 opened the pull request:
This makes it so that more of our tests get build by hydra and evaluated in our CI.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jopejoe1 committed:
tests.importCargoLock.*: remove usage of lib.fileset

The usage of lib.fileset was forbidden in nixpkgs due to a bug in nix.
For more information, see 8725e466ef2bcc5be69106c16dbf69b9ef989273 and the PR it was merged in #369694

jopejoe1 committed:
tests.config.allowPkgsInPermittedInsecurePackages: set system

The called function defaults to `builtins.currentSystem` which does not work in pure-eval

jopejoe1 committed:
tests.stdenv: remove `__attrsFailEvaluation`

These attrs do not fail to Eval on Darwin anymore

jopejoe1 committed:
tests.cross: remove `__attrsFailEvaluation`

These attrs do not currently fail eval

jopejoe1 committed:
tests: recurse into more attrs

This makes it so that our ci evals more tests and that hydra builds them, making it possible to catch more regressions.

winterqt reviewed (approved):
Rust changes look good.

How did you test this btw, just running `ci/eval`? Or did you spin up a Hydra? (Just curious.)

winterqt commented:
Looks like at least one of the Darwin tests are failing for reasons I can’t entirely grok right now on mobile — pkgconf is failing to build, that much I can see. I’ll dig into this later if you don’t by the time I get a second to.

jopejoe1 committed:
openssl: don't create `separateDebugInfo` on android

This causes an infinite recursion when evaling tests.cross.gcc.file.aarch64-android

jopejoe1 commented:
> How did you test this btw, just running `ci/eval`? Or did you spin up a Hydra? (Just curious.)

`ci/eval`, manual building of tests and a script to find wich attrsets are not being evaled.

winterqt commented:
Thanks. I'm actively working on the macOS test failures right now, getting closer and closer, just takes a while.

winterqt commented:
Alright, Darwin fixes are in #406062.

jopejoe1 committed:
tests.cross.{gcc,llvm}.*: don't recurse into

Makes the ci get killed by OOM

jopejoe1 commented:
Cleaned up the commit messages and added a bit of context to each of the commits.

philiptaron reviewed (approved):
Nice work. I reviewed it commit by commit.

trofi commented:
33964ab8ba0d110bdeeea6e51bcab74fd224990d `tests: recurse into more attrs` somehow broke `maintainers/scripts/rebuild-amount.sh` script. Now it can't diff anything at all:

```
$ ./maintainers/scripts/rebuild-amount.sh HEAD^
...
Estimating rebuild amount by counting changed Hydra jobs (parallel=unset).
error:
       … while evaluating the attribute 'tests'
       … while evaluating the attribute 'cc-wrapper'
       … while evaluating the attribute 'gccTests'
       … while evaluating the attribute 'gccMultiStdenv'
       … while evaluating the attribute 'x86_64-darwin'
...
       error: evaluation aborted with the following error message: 'cannot find attribute 'tests.cc-wrapper.gccTests.gccMultiStdenv''
```

I think it's because `tests.cc-wrapper.gccTests.gccMultiStdenv.x86_64-darwin` is subtly broken when evaluated on `x86_64-linux` host. This attribute used not to exist, but now it's incorrect.

Reproducer:

```
$ nix eval --impure --expr 'with import pkgs/top-level/release.nix {}; builtins.attrNames tests.cc-wrapper.gccTests.gccMultiStdenv.x86_64-darwin or "NONE"'
error:
       … while calling the 'attrNames' builtin
         at «string»:1:44:
            1| with import pkgs/top-level/release.nix {}; builtins.attrNames tests.cc-wrapper.gccTests.gccMultiStdenv.x86_64-darwin or "NONE"
             |                                            ^

       … while evaluating the attribute 'cc-wrapper.gccTests.gccMultiStdenv.x86_64-darwin'
         at lib/attrsets.nix:988:47:
          987|   */
          988|   nameValuePair = name: value: { inherit name value; };
             |                                               ^
          989|

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: evaluation aborted with the following error message: 'cannot find attribute 'tests.cc-wrapper.gccTests.gccMultiStdenv''
```

It used to return `"NONE"` before the change.

I think the change somehow affects `hostPlatform` used in `pkgs/test/default.nix`:

```
      gccTests =
        let
          pkgSets = lib.pipe (attrNames pkgs) (
            [
              (filter (lib.hasPrefix "gcc"))
              (filter (lib.hasSuffix "Stdenv"))
              (filter (n: n != "gccCrossLibcStdenv"))
             # ...
            ]
            ++
              lib.optionals
                (
                  !(
                    (stdenv.buildPlatform.isLinux && stdenv.buildPlatform.isx86_64)
                    && (stdenv.hostPlatform.isLinux && stdenv.hostPlatform.isx86_64)
                  )
                )
                [
                  (filter (n: !lib.hasSuffix "MultiStdenv" n))
                ]
          );
        in
        lib.genAttrs pkgSets (name: callPackage ./cc-wrapper { stdenv = pkgs.${name}; });
```

Here `(filter (n: !lib.hasSuffix "MultiStdenv" n))` is supposed to filter out the attribute on non-linux, but it does not as `stdenv.buildPlatform` is the eval platform, not the target platform. I think it's expected to be target platform, but recursion change somehow swallowed it (or exposed bad hydra job definition?).
